### DISCUSSION TITLE
[Feature] Multiple new command line options

### DISCUSSION THREAD

**Discussion by rtfmoz2** [curl/curl#9294] [Ideas]
# Summary
With some simple changes and support for variables based on the last request we can significantly increase the functionality of the command to deal with more complex use cases

## Configuration
Assume **curlrc** is set as follows
cookie-jar=cookies
cookie=cookies
no-progress-meter
 
## New variable support
Variable values are updated based current request execution. Existing commands can use var names from --write-out and headers from the previous response when using --next. The header value is the name of the header. If it does not exist returns an empty string.
curl https://www.redirectme.com --next --header "Origin: %{Access-Control-Allowed-Origin}" --url "%{redirect_url}" --header "Cookie: %{Set-Cookie}"

## New origin command 
Set the Origin header for CORS testing. This will persist globally across --next invocations until changed.
--origin null
--origin https://www.myapp.com

## New refresh option
Refresh the page for login testing, enables users to handle meta-refresh cases.
--refresh a special form of --next with a built in --url %{url}, no parameters are required.

## New redirect option
Perform a redirect based on the previous response location header. 
All the usual options can be used except for --url which this replaces when used. 
--redirect is a special version of -next with a built-in --url "%{redirect_url}"

## Examples
With the above changes curl becomes a far more powerful tool to work with.

The following request which sends 7 requests in series to simulate an OAuth login sequence ( the vars cannot be used this way presently)
curl --header "Origin: https://www.myapp.com" https://proxy.sso.com/application \\
        --next --header "Origin: null" --url "%{redirect_url}" \\
        --next --header "Origin: null" --url "%{redirect_url}" \\
        --next --header "Origin: null" --url "%{url}" \\
        --next --header "Origin: null" -d name=person -d password=xxxx -d domain=xy \\
        --next --url "%{redirect_url}"

...would be greatly simplified to...
curl --origin "https://www.myapp.com" https://proxy.sso.com/application \\
        --redirect -origin "null" # this and all following requests are Origin: null \\
        --redirect \\
        --refresh \\
        --next -d name=person -d password=xxxx -d domain=xy # POST to login \\
        --redirect

**Comment by rtfmoz2** [comment]
This issue is url's are all dynamically generated for SSO. The Origin header has to be set to test CORS but being global across all redirects means it hides the actual problem. An actual browser changes Origin to null on a redirect. I was unable to test this using curl in its present form and that drives this feature request. The additional items are just logical extensions of the current product.

**Comment by bagder** [comment]
> Existing commands can use var names from --write-out and headers from the previous response

I really like this idea!

> `--refresh` - enables users to handle meta-refresh cases

That would imply parsing and understanding content in the form of HTML to a fairly high degree. I do not anticipate that we want that.

> `--redirect` is a special version of -next with a built-in `--url`

I don't understand how that is different to `--location` ?

**Reply by rtfmoz2** [reply]
> > Existing commands can use var names from --write-out and headers from the previous response
> 
> I really like this idea!
> 
> > `--refresh` - enables users to handle meta-refresh cases
> 
> That would imply parsing and understanding content in the form of HTML to a fairly high degree. I do not anticipate that we want that.

Not at all. When curl sees this option it simply sends the same request again. Preserving all the original options, nothing  more,  It’s not possible to do this presently without ending the current request sequence.

> 
> > `--redirect` is a special version of -next with a built-in `--url`
> 
> I don't understand how that is different to `--location` ?

Location follows all redirects, this does not. This follows one redirect then moves on to the -next command. With this you can actually decide what happens next using data from the redirect response

The reason for all this is the ability to sequence requests together to reproduce a known requirement. The above enhances curls ability to do this significantly.

**Reply by jzakrzewski** [reply]
> Location follows all redirects, this does not. This follows one redirect then moves on to the -next command. With this you can actually decide what happens next using data from the redirect response

Why `--max-redirs` cannot be used for that?

**Reply by bagder** [reply]
> When curl sees this option (`--refresh`) it simply sends the same request again.

So it would basically be an alias for `--next` and all the options + URL just used for the previous transfer?

>  It’s not possible to do this presently without ending the current request sequence

So it is not an alias like I think?

**Reply by bagder** [reply]
> This follows one redirect then moves on to the next command

When do you need to follow just one redirect? Can you explain more how you anticipate such a feature to be used?

**Reply by rtfmoz2** [reply]
> > Location follows all redirects, this does not. This follows one redirect then moves on to the -next command. With this you can actually decide what happens next using data from the redirect response
> 
> Why `--max-redirs` cannot be used for that?

Well in that case, consider this an alias for --next --location --max-redirs 1 --options where options are any things we want to set for the redirect.

**Reply by rtfmoz2** [reply]
> > When curl sees this option (`--refresh`) it simply sends the same request again.
> 
> So it would basically be an alias for `--next` and all the options + URL just used for the previous transfer?
> 

You are exactly correct, without having to type all those options again this is an alias for --next including previous command options. You may even want to use --repeat instead of --refresh if you prefer a more literal interpretation. It's easier to understand from a non-html persons perspective.

**Reply by rtfmoz2** [reply]
> > This follows one redirect then moves on to the next command
> 
> When do you need to follow just one redirect? Can you explain more how you anticipate such a feature to be used?

Sure, here is a sequence that we experience for OAuth. 

- https://proxy.sso.com/application -> Access Application with no Authentication 302 Redirect to Login
- https://signon.sso.com/as/authorization.oauth2?client_id=reverse_proxy.....Start of OAuth authentication sequence -  302 Redirect
- https://signon.sso.com/as/<random_session_id>/resume.oauth2 ... Establish session using URL to track progress through an auth policy - 401 Unauthorized 
- https://signon.sso.com/as/<random_session_id>/resume.oauth2... Returns webpage with browser meta-refresh (browser calls this again) - 200 OK
- https://signon.sso.com/as/<random_session_id>/resume.oauth2 - Login Form for user authentication - 401 unauthorised now presents a login form to user.
- POST https://signon.sso.com/as/<random_session_id>/resume.oauth2 - Send login data as www-formencoded-data - 302 Redirect with authorization code back to proxy
- https://proxy.sso.com/oidc?authorization_code -> Authorization recieved, Create OIDC token and set as cookie - 302 Redirect back to original application request
- https://proxy.sso.com/application - now with Cookie can pass through proxy to application and get reply. Reply as per application now.

**Reply by jzakrzewski** [reply]
> > > Location follows all redirects, this does not. This follows one redirect then moves on to the -next command. With this you can actually decide what happens next using data from the redirect response
> > 
> > 
> > Why `--max-redirs` cannot be used for that?
> 
> Well in that case, consider this an alias for --next --location --max-redirs 1 --options where options are any things we want to set for the redirect.

I really don't buy the whole Idea of aliases. People are already lost in the plethora of curl options and switches, and you propose to add some more that should, in essence, do what is already possible but in less characters. From my perspective this will be a) only more confusing, b) a maintenance burden without clear benefit

**Reply by rtfmoz2** [reply]
We make aliases every single day. It's just a chain of actions with a name that brings them together. In this case it's a chain of options brought under one name to represent the intended effect of those actions. Thats all there is to it.

**Reply by jzakrzewski** [reply]
For my private purpose I do it too. But it does not mean there must be alias commands provided by the project. In fact this is another option to learn, that does few things, thus increasing cognitive load.
I would probably be more open to an idea, similar to what git does with aliases.

**Reply by rtfmoz2** [reply]
In that case I would be more inlined to change the behaviour of --location to a single redirect and use --follow if they want more

**Reply by bagder** [reply]
@jzakrzewski  wrote:

>  I would probably be more open to an idea, similar to what git does with aliases.

I concur. git's alias concept is very cool and powerful and this without "overloading" the tool itself with them.

**Reply by bagder** [reply]
@rtfmoz2 wrote:

> In that case I would be more inlined to change the behaviour of --location to a single redirect and use --follow if they want more

We cannot change behavior of existing options. People rely on this functionality.

**Reply by rtfmoz2** [reply]
How about we just allow users to create their own aliases. Exactly like git alias. Makes it really simple then.

**Reply by bagder** [reply]
It's nothing "just" or "simple" about that either. But sure, I would be happy to explore how that could or would work.

**Comment by rtfmoz2** [comment]
I do have a general question tho, why is this project so shy of doing basic html response interrogation. It's a very well defined standard and there is plenty of support for simple web processing, like cheerio, then people can use simple terms like this to use data from the response for queries after curl does a cheerio.load(responseBody). Maybe even include it as an option --cheerio just to trigger the support. Nothing required on your side, just tell them they can use it, provide a few examples and the rest is up to them. All you do is load support and push the response page into it. For clarity of processing to detemine when to use cherio processing use c%{code} or some other signifier to make it simple when to process the statement using cheerio engine to get the output. Then your not responsible for the internal formatting at all. If they stuff it up it because they are not using cheerio syntax properly. You don't want the burden of HTML support so this way you avoid it completely. It uses an external, well known product of versions you decide to implement. https://github.com/cheeriojs/cheerio#readme

--url c%{'form[method=POST]').prop('action')}

However its JS so alternatively look gumbo-parser(Google) and gumbo-parser-cpp if needed

**Reply by jzakrzewski** [reply]
I'd say: because curl follows the philosophy of doing one thing - data transfer. It's quite a broad thing but still only one. And it uses more protocols than just HTTP. Inspecting the actual data (which is not a part of the protocol) simply does not fit. You can always pipe the output to another utility that is designed to process it.

**Reply by rtfmoz2** [reply]
Oh I understand the rational... its just we live in a html world. Makes sense to me to have some support for querying the elements of a web page. It's not rocket science anymore. Its a very well documented standard and there is plenty of libraries out there to support inspection of html page elements. I mean thats all anyone wants. The ability to retrieve html elements from a returned page. Once library support is added then you can refer people to the library for any support.

**Reply by jzakrzewski** [reply]
> I mean thats all anyone wants. The ability to retrieve html elements from a returned page.

I wouldn't generalise if I were you. I can only speak about myself and 99% of the time I don't even have the need to look at the returned data. The remaining 1% gets pumped into other utilities for processing.

> Once library support is added then you can refer people to the library for any support.

Unfortunately, it does not work this way. I'm sure @bagder can confirm that any issue with external library first generates a ticket for the curl project which creates load for the maintainers.

**Reply by rtfmoz2** [reply]
> I wouldn't generalise if I were you

You're spot on, I shouldn't. 

> 
> Unfortunately, it does not work this way. I'm sure @bagder can confirm that any issue with external library first generates a ticket for the curl project which creates load for the maintainers.

Ahh I see. Even signposts are not effective?

**Reply by bagder** [reply]
curl doesn't parse HTML because it doesn't parse/touch/understand content at all. It has been the line in the sand for curl since before it was even named curl. :grin: 

I am always open for discussing moving such lines and even changing scope and expanding, but I find that discussing starting to parse HTML (or any other content) is a **super huge** topic and as such it requires lots of thought and discussions and it needs to be driven by some really powerful use cases.

The most important question would be: **why** would curl do this when there already exists lots of solutions for it?

**Reply by rtfmoz2** [reply]
Thank you for a such a great response @bagder 

It is a huge topic I agree. Perhaps we can approach the idea in a different way. Bear with me for a moment if you would. Essentially the requirement is to extract information from the returned data to use for the next request. I only mention html processing because your making a http request so the response has to conform to the standard in the reply however that’s headers only. The body of the response can vary widely. So rather than call our cheerio for example is their a way we can pass data and get values in return? Perhaps a concept as simple as —command that allows people to integrate their tool with curl. The concept is pass the data to the command and get a series of return values in response. Perhaps name=value separated by spaces? Then in our curl requests we can use %c{name} to retrieve the value. This lets you keep curl as it is and allows tools to provide the input required for subsequent requests

**Reply by rtfmoz2** [reply]
Before it gets asked , yes not all commands work this way. Thats why it’s the job of the person who wants to use this to write a curl-wrapper around their tool to provide a response in the format required for curl. I volunteer to write the first. But it would not be submitted here unless you want it to be @bagder . I just figure that’s up to the person using the tool.
